### Data Loading and Inspection

In [1]:
# Load dataset:
df_support <- read.csv("support2.csv", header = TRUE, sep = ",")

df_description <- read.csv("support-variables-description.csv", header = TRUE, sep = ",")


In [2]:
# df_description


In [3]:
# Inspect dataset
head(df_support)


age      death sex    hospdead slos d.time dzgroup          
1 62.84998 0     male   0         5   2029   Lung Cancer      
2 60.33899 1     female 1         4      4   Cirrhosis        
3 52.74698 1     female 0        17     47   Cirrhosis        
4 42.38498 1     female 0         3    133   Lung Cancer      
5 79.88495 0     female 0        16   2029   ARF/MOSF w/Sepsis
6 93.01599 1     male   1         4      4   Coma             
  dzclass            num.co edu ... crea      sod ph       glucose bun urine
1 Cancer             0      11  ... 1.1999512 141 7.459961 NA      NA  NA   
2 COPD/CHF/Cirrhosis 2      12  ... 5.5000000 132 7.250000 NA      NA  NA   
3 COPD/CHF/Cirrhosis 2      12  ... 2.0000000 134 7.459961 NA      NA  NA   
4 Cancer             2      11  ... 0.7999268 139       NA NA      NA  NA   
5 ARF/MOSF           1      NA  ... 0.7999268 143 7.509766 NA      NA  NA   
6 Coma               1      14  ... 0.6999512 140 7.659180 NA      NA  NA   
  adlp adls sfdm2               adlsc
1  7   7                        7    
2 NA   1    <2 mo. follow-up    1    
3  1   0    <2 mo. follow-up    0    
4  0   0    no(M2 and SIP pres) 0    
5 NA   2    no(M2 and SIP pres) 2    
6 NA   1    <2 mo. follow-up    1

In [4]:
# Check structure of dataset
str(df_support)
dim(df_support)


'data.frame':	9105 obs. of  47 variables:
 $ age     : num  62.8 60.3 52.7 42.4 79.9 ...
 $ death   : int  0 1 1 1 0 1 1 1 1 1 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: int  0 1 0 0 0 1 0 0 0 0 ...
 $ slos    : int  5 4 17 3 16 4 9 7 12 8 ...
 $ d.time  : int  2029 4 47 133 2029 4 659 142 63 370 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ num.co  : int  0 2 2 2 1 1 1 3 2 0 ...
 $ edu     : int  11 12 12 11 NA 14 14 NA 12 11 ...
 $ income  : Factor w/ 5 levels "","$11-$25k",..: 2 2 5 5 1 1 3 1 1 3 ...
 $ scoma   : int  0 44 0 0 26 55 0 26 26 0 ...
 $ charges : num  9715 34496 41094 3075 50127 ...
 $ totcst  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ totmcst : num  NA NA NA NA NA NA NA NA NA NA ...
 $ avtisst : num  7 29 13 7 18.7 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 4 ...
 $ sps

[1] 9105   47

After Inspecting closely, We see that for many categorical variables', the datatype are not 'Factor' eg. $ sex : chr  "male" "female" etc.

### Data Cleaning and Preprocessing

In [5]:
# Identify and Group the categorical variables as categories of interest
cat_interest <- c("death", "sex", "hospdead", "dzgroup", "dzclass", "income", "race", "diabetes", "dementia", "ca", "dnr", "adlp", "adls", "sfdm2")

# Apply unique function to check unique categories for each variable and check their structure
unique_values <- sapply(df_support[cat_interest], unique)
unique_values

str(unique_values)


$death
[1] 0 1

$sex
[1] male   female
Levels: female male

$hospdead
[1] 0 1

$dzgroup
[1] Lung Cancer       Cirrhosis         ARF/MOSF w/Sepsis Coma             
[5] CHF               Colon Cancer      COPD              MOSF w/Malig     
8 Levels: ARF/MOSF w/Sepsis CHF Cirrhosis Colon Cancer Coma ... MOSF w/Malig

$dzclass
[1] Cancer             COPD/CHF/Cirrhosis ARF/MOSF           Coma              
Levels: ARF/MOSF Cancer Coma COPD/CHF/Cirrhosis

$income
[1] $11-$25k   under $11k            $25-$50k   >$50k     
Levels:  $11-$25k $25-$50k >$50k under $11k

$race
[1] other    white    black    hispanic asian            
Levels:  asian black hispanic other white

$diabetes
[1] 0 1

$dementia
[1] 0 1

$ca
[1] metastatic no         yes       
Levels: metastatic no yes

$dnr
[1] no dnr                          dnr after sadm  dnr before sadm
Levels:  dnr after sadm dnr before sadm no dnr

$adlp
[1]  7 NA  1  0  2  3  5  6  4

$adls
[1]  7  1  0  2 NA  5  4  6  3

$sfdm2
[1]                     <2 mo. follow-up    no(M2 and SIP pres)
[4] SIP>=30             adl>=4 (>=5 if sur) Coma or Intub      
6 Levels:  <2 mo. follow-up adl>=4 (>=5 if sur) ... SIP>=30

List of 14
 $ death   : int [1:2] 0 1
 $ sex     : Factor w/ 2 levels "female","male": 2 1
 $ hospdead: int [1:2] 0 1
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 1 5 2 4 6 8
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 1 3
 $ income  : Factor w/ 5 levels "","$11-$25k",..: 2 5 1 3 4
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 3 4 2 1
 $ diabetes: int [1:2] 0 1
 $ dementia: int [1:2] 0 1
 $ ca      : Factor w/ 3 levels "metastatic","no",..: 1 2 3
 $ dnr     : Factor w/ 4 levels "","dnr after sadm",..: 4 1 2 3
 $ adlp    : int [1:9] 7 NA 1 0 2 3 5 6 4
 $ adls    : int [1:9] 7 1 0 2 NA 5 4 6 3
 $ sfdm2   : Factor w/ 6 levels "","<2 mo. follow-up",..: 1 2 5 6 3 4


In [6]:
# convert categorical and ordinal variables to factors  # nolint
df_support$sex <- factor(df_support$sex)

df_support$death <- factor(df_support$death)

df_support$hospdead <- factor(df_support$hospdead)

df_support$dzgroup <- factor(df_support$dzgroup)

df_support$dzclass <- factor(df_support$dzclass)

df_support$income <- factor(df_support$income, ordered = TRUE, levels = c("under $11k", "$11-$25k", "$25-$50k", ">$50k"))

df_support$race <- factor(df_support$race)

df_support$diabetes <- factor(df_support$diabetes)

df_support$dementia <- factor(df_support$dementia)

factor(c("metastatic", "no", "yes"),
    levels = c("no", "yes", "metastatic"),
    ordered = TRUE
)

df_support$adlp <- factor(df_support$adlp, ordered = TRUE, levels = c(1, 2, 3, 4, 5, 6, 7))

df_support$adls <- factor(df_support$adls, ordered = TRUE, levels = c(1, 2, 3, 4, 5, 6, 7))

df_support$sfdm2 <- factor(df_support$sfdm2, ordered = TRUE, levels = c("no(M2 and SIP pres)", "adl>=4 (>=5 if sur)", "SIP>=30", "Coma or Intub", "<2 mo. follow-up"))


[1] metastatic no         yes       
Levels: no < yes < metastatic

In [7]:
# Check structure of dataset to confirm changes
str(df_support[cat_interest])


'data.frame':	9105 obs. of  14 variables:
 $ death   : Factor w/ 2 levels "0","1": 1 2 2 2 1 2 2 2 2 2 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: Factor w/ 2 levels "0","1": 1 2 1 1 1 2 1 1 1 1 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ income  : Ord.factor w/ 4 levels "under $11k"<"$11-$25k"<..: 2 2 1 1 NA NA 3 NA NA 3 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 4 ...
 $ diabetes: Factor w/ 2 levels "0","1": 1 1 1 1 1 1 1 2 1 1 ...
 $ dementia: Factor w/ 2 levels "0","1": 1 1 1 1 1 1 1 1 2 1 ...
 $ ca      : Factor w/ 3 levels "metastatic","no",..: 1 2 2 1 2 2 2 2 1 1 ...
 $ dnr     : Factor w/ 4 levels "","dnr after sadm",..: 4 1 4 4 4 4 4 4 2 4 ...
 $ adlp    : Ord.factor w/ 7 levels "1"<"2"<"3"<"4"<..: 7 NA 1 NA NA NA NA NA NA NA ...
 $ adls    : Ord.factor w/ 7 levels "1"<"2"<"3"<

### MISSING VALUES ANALYSIS



In [8]:
# Count the total number of complete cases (Rows with no missing values)
num_complete_cases <- sum(complete.cases(df_support))
num_complete_cases

dim(df_support)

# We have only 106 complete cases (rows with no missing values) out of 9105 rows. This is a very small number of complete cases. In order to make the dataset usable, we will have to find identify which variables have the most missing values and drop them from the dataset.


[1] 106

[1] 9105   47

In [9]:
# Check for Missing Values in each column and sort in descending order

na_counts_col <- colSums(is.na(df_support))

na_counts_col <- sort(na_counts_col, decreasing = TRUE)

na_counts_col[1:15]


adlp    adls   urine glucose     bun totmcst     alb  income    bili    pafi 
   7490    5975    4862    4500    4352    3475    3372    2982    2601    2325 
     ph   prg2m     edu   prg6m   sfdm2 
   2284    1649    1634    1633    1400

Variables with the most missing values are:
1. adlp : 7490 (Too many missing values, and highly correlated with other 'Activities of Daily Living Index' variable (adlsc) 
2. adls : 7490 (same as above)
3. urine : 4862
4. glucose : 4500
5. bun : 4352
6. totmcst (Total micro cost): 3475 (Highly correlated with totcst)
7. alb : 3372  etc.

To determine which variables to drop, we will create a correlation matrix and identify the variables with high correlation. 



To start, We will drop the variables with more than 50% missing values. These are:
1. adlp and 
2.  

In [10]:
# Create df with only numeric columns for creating a correlation matrix
numeric_columns <- sapply(df_support, is.numeric)
numeric_df <- df_support[numeric_columns]

head(numeric_df)


age      slos d.time num.co edu scoma charges totcst totmcst avtisst  ...
1 62.84998  5   2029   0      11   0     9715   NA     NA       7.00000 ...
2 60.33899  4      4   2      12  44    34496   NA     NA      29.00000 ...
3 52.74698 17     47   2      12   0    41094   NA     NA      13.00000 ...
4 42.38498  3    133   2      11   0     3075   NA     NA       7.00000 ...
5 79.88495 16   2029   1      NA  26    50127   NA     NA      18.66666 ...
6 93.01599  4      4   1      14  55     6884   NA     NA       5.00000 ...
  pafi     alb      bili      crea      sod ph       glucose bun urine adlsc
1 388.0000 1.799805 0.1999817 1.1999512 141 7.459961 NA      NA  NA    7    
2  98.0000       NA        NA 5.5000000 132 7.250000 NA      NA  NA    1    
3 231.6562       NA 2.1997070 2.0000000 134 7.459961 NA      NA  NA    0    
4       NA       NA        NA 0.7999268 139       NA NA      NA  NA    0    
5 173.3125       NA        NA 0.7999268 143 7.509766 NA      NA  NA    2    
6 266.6250       NA        NA 0.6999512 140 7.659180 NA      NA  NA    1

In [11]:
# Create a corrplot to visualize the correlation between the variables
# install.packages("corrplot")
library(corrplot)
library(reshape2)

# Assuming df is your data frame
correlation_matrix <- cor(numeric_df, use = "complete.obs")

# Flatten and filter
corr_melted <- melt(correlation_matrix)

# Remove self-correlations and duplicates
corr_melted <- subset(corr_melted, Var1 != Var2)
corr_melted <- corr_melted[!duplicated(t(apply(corr_melted[, 1:2], 1, sort))), ]

# Sort by absolute correlation value
corr_melted <- corr_melted[order(-abs(corr_melted$value)), ]
top_correlations <- head(corr_melted, 100) # Top 20 correlations

print(top_correlations)


corrplot 0.92 loaded


        Var1    Var2      value
240  totmcst  totcst  0.9689325
410   surv6m  surv2m  0.9603923
51    dnrday    slos  0.8952043
512    prg6m   prg2m  0.8948647
41    totcst    slos  0.8333012
206   totcst charges  0.8328858
42   totmcst    slos  0.8173748
207  totmcst charges  0.8080156
249   dnrday  totcst  0.7836419
282   dnrday totmcst  0.7706970
342      aps     sps  0.7701623
343   surv2m     sps -0.7460156
889      bun    crea  0.7023728
344   surv6m     sps -0.6695264
40   charges    slos  0.6188178
376   surv2m     aps -0.5969726
309      aps avtisst  0.5939339
216   dnrday charges  0.5926286
274  avtisst totmcst  0.5624625
213     hday charges  0.5623091
412    prg2m  surv2m  0.5547620
178   surv2m   scoma -0.5482460
241  avtisst  totcst  0.5477100
446    prg6m  surv6m  0.5396185
308      sps avtisst  0.5265082
445    prg2m  surv6m  0.5264234
413    prg6m  surv2m  0.5120520
377   surv6m     aps -0.4926692
208  avtisst charges  0.4907841
83     prg6m  d.time  0.4863762
80    su

In [12]:
# # Assuming x and y are your variables
# result <- cor.test(df_support$dnrday, df_support$avtisst, method = "pearson")

# # This will give you the correlation coefficient and the p-value
# print(result)


In [13]:
# Drop variables with high missing values from the list

# And 'column_to_drop' is the name of the column you want to drop
df_updated_mis <- subset(df_support, select = -c(adlp, adls, urine, totmcst, glucose))

# Check structure of dataset to confirm changes
str(df_updated_mis)


'data.frame':	9105 obs. of  42 variables:
 $ age     : num  62.8 60.3 52.7 42.4 79.9 ...
 $ death   : Factor w/ 2 levels "0","1": 1 2 2 2 1 2 2 2 2 2 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: Factor w/ 2 levels "0","1": 1 2 1 1 1 2 1 1 1 1 ...
 $ slos    : int  5 4 17 3 16 4 9 7 12 8 ...
 $ d.time  : int  2029 4 47 133 2029 4 659 142 63 370 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ num.co  : int  0 2 2 2 1 1 1 3 2 0 ...
 $ edu     : int  11 12 12 11 NA 14 14 NA 12 11 ...
 $ income  : Ord.factor w/ 4 levels "under $11k"<"$11-$25k"<..: 2 2 1 1 NA NA 3 NA NA 3 ...
 $ scoma   : int  0 44 0 0 26 55 0 26 26 0 ...
 $ charges : num  9715 34496 41094 3075 50127 ...
 $ totcst  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ avtisst : num  7 29 13 7 18.7 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 

In [14]:
# Recheck the total number of complete cases (Rows with no missing values) after dropping variables with high missing values
num_complete_cases <- sum(complete.cases(df_updated_mis))
num_complete_cases


[1] 922

In [15]:
# Lets do more indepth missing value analysis using the 'mice' package
# install.packages("lifecycle")
# install.packages("mice")

library(mice)

# Plot the missing values using the 'md.pattern' function
md.pattern(df_updated_mis)


Warning message:
"package 'mice' was built under R version 3.6.3"

ERROR: Error: package or namespace load failed for 'mice' in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]):
 there is no package called 'lifecycle'
